### PR TITLE
[ContactStructuralMechanicsApplication] WIP Mortar penalty condition for contact

### PR DESCRIPTION
**Pull request by loumalouomega**
The full support for explicit integration scheme will be provide in a future PR (I dont want to make the PR huge). The frictional formulation will be added in a future PR, but this one already integrates many utilities related with the frictional formulation

This is a big PR and it includes #3787

### CODE REVIEW

File: applications/ContactStructuralMechanicsApplication/custom_conditions/mortar_contact_condition.cpp [hunk 1]
```diff
@@ -0,0 +1,1014 @@
+// KRATOS  ___|  |       |       |
```
Comment (loumalouomega) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/mortar_contact_condition.cpp:1]: This is not detected as a rename for some reason, but it is basically the old ALM_mortar_contact_condition.cpp with some duplicated code removed (that's why is 800 lines lighter)

File: applications/ContactStructuralMechanicsApplication/custom_conditions/mortar_contact_condition.h [hunk 2]
```diff
@@ -9,8 +9,8 @@
 //  Main authors:  Vicente Mataix Ferrandiz
 //
```
Comment (loumalouomega) [inline applications/ContactStructuralMechanicsApplication/custom_conditions/mortar_contact_condition.h:11]: And in this case the rename is detected for the header¿?

**Comment by philbucher** [comment]
I will review after you are done with your commits
Please let me know

**Comment by loumalouomega** [comment]
In this other case I found a 🐛, I will notice you

**Comment by loumalouomega** [comment]
I think this is all

File: applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    "restart_options"          : {
+        "SaveRestart"      : false,
+        "RestartFrequency" : 0,
+        "LoadRestart"      : false,
+        "Restart_Step"     : 0
+    },
+    "constraints_data"         : {
+        "incremental_load"         : false,
+        "incremental_displacement" : false
+    }
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json]: outdated settings

File: applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    "_json_output_process" : [
+    {
+        "python_module"   : "json_output_process",
+        "kratos_module" : "KratosMultiphysics",
+        "help"                  : "",
+        "process_name"          : "JsonOutputProcess",
+        "Parameters"            : {
+            "output_variables"     : ["DISPLACEMENT_Z"],
+            "output_file_name"     : "penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_results.json",
+            "model_part_name"      : "Contact_Part",
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json]: missing main-model-part-name

File: applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_materials.json [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                    "Variables": {
+                            "DENSITY": 1.00000E+03,
+                            "YOUNG_MODULUS": 2.06900E+11,
+                            "POISSON_RATIO": 2.90000E-01
+                    },
+                    "Tables": {}
+            }
+    },
+    {
+            "model_part_name": "Parts_Parts_Auto2",
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_materials.json:20]: missing main-model-part-name

File: applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/displacement_lagrangemultiplier_contact_criteria.h [hunk 6]
```diff
@@ -150,7 +150,7 @@ class DisplacementLagrangeMultiplierContactCriteria
         mLMRatioTolerance =  ThisParameters["contact_displacement_relative_tolerance"].GetDouble();
         mLMAbsTolerance =  ThisParameters["contact_displacement_absolute_tolerance"].GetDouble();
 
-        // Additional flags -> NOTE: Replace for a ral flag?¿
+        // Additional flags -> NOTE: Replace for a real flag?¿
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/displacement_lagrangemultiplier_contact_criteria.h]: strange character needed?
```suggestion
        // Additional flags -> NOTE: Replace for a real flag?
```

File: applications/ContactStructuralMechanicsApplication/python_scripts/penalty_contact_process.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+import KratosMultiphysics.ContactStructuralMechanicsApplication as CSMA
+
+def Factory(settings, Model):
+    if(type(settings) != KM.Parameters):
+        raise Exception("Expected input shall be a Parameters object, encapsulating a json string")
+    return PenaltyContactProcess(Model, settings["Parameters"])
+
+import sys
+
+import alm_contact_process
```
Comment (philbucher) [inline applications/ContactStructuralMechanicsApplication/python_scripts/penalty_contact_process.py]:
```suggestion
import KratosMultiphysics.ContactStructuralMechanicsApplication.alm_contact_process as alm_contact_process
```

Comment (philbucher) [reply applications/ContactStructuralMechanicsApplication/python_scripts/penalty_contact_process.py]: better to use the new syntax from now on

**Review by philbucher — DISMISSED** [review]
Take a look at codacy. Esp the unised imports and the suggestions related to `enumerate` (Not too long ago I discovered how to properly use `enumerate` and `zip` => game changer, worth looking it up if you don't know it already!)

other than that look sok to me, even though I cannot understand in detail what is happening

File: applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/displacement_lagrangemultiplier_contact_criteria.h (hunk 6 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/custom_strategies/custom_convergencecriterias/displacement_lagrangemultiplier_contact_criteria.h]: I would say non ASCII character ¿ is not strange In Spanish, you needed to start a question (which is very practical when reading a long question in my opinion)

File: applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_materials.json (hunk 5 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_materials.json:20]: OK

File: applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json (hunk 4 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json]: OK

File: applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json (hunk 3 above)
Comment (loumalouomega) [reply applications/ContactStructuralMechanicsApplication/tests/penalty_frictionless_contact_test_3D/3D_contact_simplest_patch_matching_test_parameters.json]: Copy pasted, sorry

**Comment by loumalouomega** [comment]
I will tackle the rest of your comments in a moment

**Comment by loumalouomega** [comment]
@philbucher I applied your suggestions before checking the Codacy and now I don't know what is wrong, I will address that in a future PR, what do you mean by the enum and zip?

**Comment by philbucher** [comment]
> @philbucher I applied your suggestions before checking the Codacy and now I don't know what is wrong, I will address that in a future PR, what do you mean by the enum and zip?

[enumerate](http://book.pythontips.com/en/latest/enumerate.html)
[zip](https://www.programiz.com/python-programming/methods/built-in/zip)

**Review by philbucher — APPROVED** [review]

**Comment by loumalouomega** [comment]
Thanks

**Comment by loumalouomega** [comment]
> > @philbucher I applied your suggestions before checking the Codacy and now I don't know what is wrong, I will address that in a future PR, what do you mean by the enum and zip?
> 
> [enumerate](http://book.pythontips.com/en/latest/enumerate.html)
> [zip](https://www.programiz.com/python-programming/methods/built-in/zip)

I see, but I think in this case zip fits better, but good to know :D

**Pull request merged**